docs/java.md

```diff
@@ -0,0 +1,373 @@
+
+```java
+    Value getHippoResults = client.query(
+        Select(Value("data"),Get(hippoRef))
+    ).get();
+    System.out.println("Hippo Spells:\n " + toPrettyJson(getHippoResults) + "\n");
+```
+
+#### How to retrieve the values from a query result
+
+That query returns the data in the form of a json object.  The data can be extracted from the results by using:
+
+```java
+    //convert the hippo results into primitive elements
+    String element = getHippoResults.get(Field.at("element").to(String.class));
+    System.out.println("spell element = " + element);
+```
+
+Later on we will show a better method that uses User Defined types to transform this automatically
+
+#### How to safely work with result objects
+
+This object represents the result of an operation and it might be success or a failure. All conversion operations returns an object like this. This way it's possible to avoid check for nullability everywhere in the code.
+
+```java
+    //This would return an empty option if the field is not found or the conversion fails
+    Optional<String> optSpellElement = getHippoResults.getOptional(Field.at("element").to(String.class));
+    if (optSpellElement.isPresent()) {
+        String element2 = optSpellElement.get();
+        System.out.println("optional spell element 2 = " + element2);
+    }
+    else {
+        System.out.println("Something went wrong reading the spell");
+    }
+```
+
+Optionally it's possible transform one `Result<T>` into another `Result<T>` of different type using `map` and `flatMap`.  If the `result` represents an failure all calls to `map` and `flatMap` are ignored and it returns a new failure with the same error message. See `com.faunadb.client.types.Result` for details.
+
+
+#### How to execute a list query and retrieve a collection of the results
+
+The `query` method takes an `Expr` object. `Expr` objects can be composed with others `Expr` to create complex query objects. `com.faunadb.client.query.Language` is a helper class where you can find all available expressions in the library.
+
+```java
+    Value queryIndexResults = client.query(
+        SelectAll(Path("data", "id"),
+            Paginate(
+                Match(Index(Value(INDEX_NAME)))
+            ))
+    ).get();
+```
+
+That query returns a list of resource references to all the spells in the index.  The list of references can be extracted from the results by using:
+
+```java
+    Collection<String>  spellsRefIds = queryIndexResults.asCollectionOf(String.class).get();
+    System.out.println("spellsRefIds = " + spellsRefIds);
+```
+
+
+### How to work with user defined classes
+
+Instead of manually creating your objects via the DSL (e.g. the ObjectV), you can use annotations to automatically encode and decode the class to user-defined types.  These transform the types into the equivalent `Value` types.
```

Comment:
pls, do not mention `ObjectV` as if the user could create instances of that class, instead mention the constructor classes like `Obj()` or `Arr()`